REGRESSION(299900@main): Crash in Style::Builder::ensureRollbackCascadeForRevert() with user stylesheet

anttijk opened the pull request:
#### bbf8d7c49cc111a761a9365b53daec5cfece2bee
<pre>
REGRESSION(299900@main): Crash in Style::Builder::ensureRollbackCascadeForRevert() with user stylesheet
<a href="https://bugs.webkit.org/show_bug.cgi?id=302386">https://bugs.webkit.org/show_bug.cgi?id=302386</a>
<a href="https://rdar.apple.com/164300744">rdar://164300744</a>

Reviewed by Kimmo Kinnunen.

Test: fast/css/user-stylesheet-rollback.html
* LayoutTests/fast/css/user-stylesheet-rollback-expected.txt: Added.
* LayoutTests/fast/css/user-stylesheet-rollback.html: Added.
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::makeRollbackCascadeKey):

299900@main changes enums so that value PropertyCascade::Origin::UserAgent is zero.
This caused RollbackCascadeKey&apos;s empty value (all zeros) collide with a legal value.
Fix by adding an extra bool to the key which is always set on legal values.

* Source/WebCore/style/StyleBuilder.h:

Canonical link: <a href="https://commits.webkit.org/302903@main">https://commits.webkit.org/302903@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1b8544c6fbd91f9964518609839e1096fa5fdade

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130617 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2888 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41572 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138037 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/82237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/170494ed-e3d2-4c8b-aeb0-64862015f2bd) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/132488 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2900 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2780 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99510 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/82237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/31f32a44-dd17-4ff8-b82b-5dac3fac9e33) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133564 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/2102 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116958 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80217 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8c650996-3c24-4189-b371-d88eb35c508b) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/2023 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35092 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81294 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110595 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35594 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140516 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2678 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2429 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108014 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2722 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113302 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107946 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27464 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2065 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31729 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55656 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2748 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66137 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2567 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2769 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2674 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebCore/style/StyleBuilder.cpp:
```diff
@@ -756,7 +756,8 @@ const PropertyCascade* Builder::ensureRollbackCascadeForRevertLayer()
 {
-    return { static_cast<unsigned>(cascadeOrigin), static_cast<unsigned>(scopeOrdinal), static_cast<unsigned>(cascadeLayerPriority) };
+    static constexpr auto isNonEmptyValue = true;
+    return { static_cast<unsigned>(cascadeOrigin), static_cast<unsigned>(scopeOrdinal), static_cast<unsigned>(cascadeLayerPriority), isNonEmptyValue };
```
What about:
```
// Use + 1 in order to ensure 0,0,0,0 is reserved for empty value.
....
static_cast<unsigned>(cascadeOrigin) + 1
```

anttijk replied:
I like this pattern better (it is used elsewhere too) as it is a universal solution.

anttijk committed:
REGRESSION(299900@main): Crash in Style::Builder::ensureRollbackCascadeForRevert() with user stylesheet
https://bugs.webkit.org/show_bug.cgi?id=302386
rdar://164300744

Reviewed by Kimmo Kinnunen.

Test: fast/css/user-stylesheet-rollback.html
* LayoutTests/fast/css/user-stylesheet-rollback-expected.txt: Added.
* LayoutTests/fast/css/user-stylesheet-rollback.html: Added.
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::makeRollbackCascadeKey):

299900@main changes enums so that value PropertyCascade::Origin::UserAgent is zero.
This caused RollbackCascadeKey's empty value (all zeros) collide with a legal value.
Fix by adding an extra bool to the key which is always set on legal values.

* Source/WebCore/style/StyleBuilder.h:

Canonical link: https://commits.webkit.org/302903@main
